### ISSUE TITLE
Trace to memory buffer.

### ISSUE THREAD

**Issue by vext01** [softdevteam/hwtracer#8]
Currently the trace goes to a file. Eventually, once we understand more, we will need to trace to a buffer and decode it in parallel with the interpreter.

**Comment by vext01** [comment]
This is one of the next tasks on my list.

I think the tracer backends should include a way to recieve "chunks" of the trace incrementally, probably via a channel. This should be built into the `Tracer` trait so that it can be done in a generic fashion for all backends.

**Comment by vext01** [comment]
This is next on the list.

The code that reads trace data out of the `mmap`ed buffer is currently written in C, and runs in a thread:
https://github.com/softdevteam/traceme/blob/a932bec22f7e59495d770f59b409b9bdcc7cf799/src/backends/perf_pt/perf_pt.c#L195

We need to think about how to get this out of C, and into rust.

Goodish Ideas:
 1. Pipe 
 2. Have the C code heap allocate and have an accessor method to get at the raw pointer once we are done tracing.
 3. Try to translate the poll loop into Rust code and use a channel.

`1` would work, but might introduce delays.

`2` is good, but would disallow "streaming" the trace. We might like to process the trace in parallel with it being collected?

`3` is good, but would require some (hopefully fairly light) architectural changes (spawn the thread in Rust, not C).

I'm tempted to try 3.

Do you have any thoughts @ltratt 

[Bad ideas i'm not conisidering are: named pipes, files, shared memory, mongodb^W :smile: )

**Comment by ltratt** [comment]
Quick thoughts: no way to pipes (*massive* overheads, relatively speaking); I don't mind 2 or 3 too much. I'm not sure we're going to be able to stream the data in practise, and certainly not at first. So if 3 is very hard, I'm OK with 2.

**Comment by vext01** [comment]
OK, let me expand on `2` a moment then.

What I'd do is start by allocating a buffer as big as the trace buffer itself, and empty the contents of the PT buffer as the poll loop is awoken. If the buffer becomes full, I'd probably reallocate it at twice the capacity and continue.

Once the trace is finished, we'd need a `tracer.trace_buf() -> *const c_void` or similar method which returns the address of the buffer. This will eventually be sent off to a packet decoder (probably a binding to libipt (https://github.com/01org/processor-trace) so we can reconstruct the instruction stream.

`2` is easier than `3`. The only nit is that we will have to provide a method to free the trace buffer once we are done.

Still OK for me to go ahead with `2`?

**Comment by ltratt** [comment]
For the time being, I think just allocating a buffer and assuming everything will fit in it is OK. We can get clever with resizing and retrying later. [Besides, whatever happens, wecould never guarantee that e.g. a second thread could read the data out of the buffer faster than it's being dumped in.]

**Comment by vext01** [comment]
I'm OK with static allocation as long as we crash loudly if the buffer gets full.

**Comment by ltratt** [comment]
I think you can take as a given that I was not thinking of silent failure ;)

**Comment by vext01** [comment]
Best to be sure, sir :)

**Comment by vext01** [comment]
We need to think a bit about the API used to expose the trace buffer.

Currently I have:
```rust
pub trait Tracer {
    ...
    /// Get the contents of the trace buffer.    
    ///     
    /// Returns a pair, whose first element is the buffer itself (as a raw pointer), and whose
    /// second element is the length of the trace.                                 
    ///                                                                            
    /// [start_tracing](trait.Tracer.html#method.start_tracing) and                
    /// [stop_tracing](trait.Tracer.html#method.stop_tracing) must have been called prior.
    fn trace_buf<'tb>(&self) -> (*const c_void, u32);
}
```

I'm thinking about the return type of this function.

Is this library only going to be used to Intel PT? If so, it's probably OK to expose raw pointers, as the trace will always be a stream of bytes (representing PT packets).

If this library will be used for other kinds of tracer, then perhaps the trace can be delivered in different formats? What about the (planned?) software version of our meta-tracer? Is traceme going to cater for this? Under that scenario, I imagine our own traces would be delivered back as in-memory data structures, rather than as a byte stream. If, on the other hand, the software tracer will use another library, then we can say that traceme is specific to Intel PT, which simplifies things a bit.

Any ideas?

[I think with a bit more fiddling, I might be able to expose the buffer as `&[u8]` (an array of bytes) instead of as a pointer, but this involves wrestling with the borrow checker as it's oblivious to buffers allocated in the FFI. Bear with me on that.]

**Comment by ltratt** [comment]
I don't think we have to overthink this API, because it's going to be internal to our overall system (i.e. interpreter authors are never going to see it). So we can start with something simple and, if it needs generalising, we can probably adjust it without killing ourselves.

That said, it occurs to me that, for our current purposes, all we care about is the stream of jump positions. So one possibility might be to expose an iterator that simply generates addresses (as `usize` or `ptr`? dunno...) captured by the trace, leaving it to the consumer to do something with them (in our case translate them into MIR blocks).

**Comment by vext01** [comment]
OK. So I'll assume that we only care about PT in traceme.

> That said, it occurs to me that, for our current purposes, all we care about is the stream of jump positions.

I'm a bit far off that yet. We need to decode the packets and reconstruct the instruction stream. This will involve bindings to libipt from Intel.

If we are going to be flexible, then we can expose the buffer for now, and later expose the iterator instead?

**Comment by ltratt** [comment]
I'm not sure we can do anything with the buffer that doesn't involve decoding the packets? So maybe the short-term fix is to have an iterator which exposes a sequence of decoded things? That at least would keep the dependency on libipt in this library?

**Comment by vext01** [comment]
> I'm not sure we can do anything with the buffer that doesn't involve decoding the packets?

Correct.

> So maybe the short-term fix is to have an iterator which exposes a sequence of decoded things?

Yep, but I'd like that to be a separate PR, and I want to fix #23 before things get too baked in.

So I think for this PR is to:
 * Not expose the buffer at all for now?
 * Have my tests look inside at the buffer.
 * Have a debug facility to dump traces to file under well formed (but not configurable) filenames, so I can test if what's coming out of the buffer is making sense.

Sounds good?

**Comment by ltratt** [comment]
Question: is the `Tracer` trait the thing that *does* tracing, or does it *store* a trace?

If the former, I have a cunning plan. Expose the buffer as an opaque struct but without any methods. Then external libraries can hold a reference to a trace, even if they can't currently do anything with it. We can add methods to do things like iterating over decoded packets on that struct later.

Apart from that, your suggestions sounds good.

**Comment by vext01** [comment]
> Question: is the Tracer trait the thing that does tracing, or does it store a trace?

At the moment it does both. But the buffer itself is born and written in C code.

> Then external libraries can hold a reference to a trace, even if they can't currently do anything with it.

Not sure I follow. But maybe you envisage that the trace decoding is done in another library outside of traceme? To be honest, if traceme is going to PT specific, we might as well do the decoding here and keep the buffer internal.

If someone finds a need to get at the trace buffer directly, we can open it up for them later.

**Comment by ltratt** [comment]
If you think about it, what we want is to get the trace out of the tracer: we will want to store them separately. So we need to separate the concepts of "tracer" and "trace", if not now, then in the nearish future. I'd be inclined to do it now, as it'll affect all your tests and so on.

I might suggest something like:

```rust
trait Tracer {
    fn trace_code(...) -> Box<Trace>
}

trait Trace {
}

struct PTTrace;

impl Trace for PTTrace {
}
```

One question is what arguments `trace_code` should take. A function or ... ?

**Comment by vext01** [comment]
What is `Trace` here? A PT packet stream, or a trace that can be executed in a JIT?

Starting to wonder if we should not call the PT-level trace a "trace", but instead a "packet stream".

**Comment by vext01** [comment]
I think it's the former, right?

**Comment by ltratt** [comment]
"Packet stream" is a meaningless term (and long) IMHO. We're clearly talking about a trace, but it's a low-level trace. We could call it BBTrace (Basic Block Trace)?

**Comment by vext01** [comment]
> "Packet stream" is a meaningless term (and long) IMHO

For better or worse, the output of PT is a collection of PT packets.

> We could call it BBTrace (Basic Block Trace)?

I would accept `PTTrace` then it's clear. PT packets do not directly correlate to basic blocks.

**Comment by vext01** [comment]
I'd even prefer `PTPackets` then there's zero chance of confusion.

**Comment by ltratt** [comment]
OK, we should take a step back. Is this crate just for processor tracing? Or is the type it exposes supposed to capture software tracing too?

**Comment by vext01** [comment]
Right! Exactly! This is what I was asking earlier:

> Is this library only going to be used to Intel PT?
> ...
> If this library will be used for other kinds of tracer, then perhaps the trace can be delivered in different formats? What about the (planned?) software version of our meta-tracer? Is traceme going to cater for this?

I'm thinking that the software tracer necessarily have will be specific to yorick, so if we put yorick code into traceme, then traceme can no longer be used for general traceing of a Rust program.

Why don't we say that traceme is a Rust library for getting the instruction stream of any given thread using Intel PT? As you've said above, the trace packets themselves are not useful, so I don't see any real point in exposing packets.

If you agree, then can we talk about the interface again.

**Comment by vext01** [comment]
And sorry this is dragging, but we have to get it right.

**Comment by ltratt** [comment]
Agree: making this a PT library makes sense. We should probably call it `pttrace`?!

On that basis, I think I'm inclined to suggest the structure I posted above of Tracer/PTTrace and so on. I'd prefer not to use "packets", which, from an external POV, seems to me an incidental detail? [We could potentially use this crate for non-Intel PT, which may not use the "packets" terminology.]

I'm also fine with getting this right, so no worries there!

**Comment by vext01** [comment]
OK good, so this is a processor trace library, but not necessarily one specific to Intel's implementation. We can rename the repo if you like, yes. Perhaps not `pttrace` (processor trace trace?).

Can you tell me what `trace_code` does in the interface you proposed?

**Comment by ltratt** [comment]
`PTrace` is also a bit ambiguous! How about `HWTrace`?

I thought of `trace_code` as "start tracing code starting at position X until it goes back to X (or some other condition is reached, but that's a detail for later I think) and return a trace object".

**Comment by vext01** [comment]
I see.

> start tracing code starting at position X until it goes back to X

That's not quite how I expected it to work. Right now we have explicit start/stop tracing calls in the library, and you don't pass any addressed. You can think of it as "start tracing now" and "stop tracing now". You should have seen these calls in the library already:

```
https://github.com/softdevteam/traceme/blob/7d2adf27bda255a3045b9760c58945ab5dffbc13/src/lib.rs#L55-L62
```

I'm not sure if we can do "start tracing code starting at position X until it goes back to X". I think there's a mode for "trace between these two addresses", but Mario said he couldn't get that one working. Also, we would need to constantly re-program the start/stop address every time there's a call, right?

I'm more interested in the interface to the instruction stream. A general interface might be an iterator over assembler instructions, but for our purposes we only need to know where there's a jump, right?. We would still need to decode the instruction stream since the direct jumps wont appear in the PT packet stream.

So how about a function `fn jumps()` which returns an iterator over some kind of a `Jump`, where we have something like:
```
enum Jump {
    Unconditional(addr), // e.g. JMP 0xdeadbeef
    Conditional(addr),  // e.g. JG 0xdeadbeef
    Indirect(addr), // e.g. J [eax]
}
```

We might need to iterate on this, but that won't come until a later PR anyway.

**Comment by ltratt** [comment]
I've just realised a possible problem with the `start`/`stop` interface: what happens in the presence of threads? Or can you have multiple `Tracer` objects, at which point its basically the caller's job to tie the `start`/`stop` calls to a single thread?

**Comment by vext01** [comment]
Each tracer instance traces one thread. Perf deals with this for us.

You  merged a fix for this earlier, remember? target_tid :)

**Comment by ltratt** [comment]
OK, so then should we just have a `Trace` struct on which there's `start_record` (which wipes whatever trace is inside) and `stop`? It's a subtle change from `Tracer`, I admit!

[I appreciate we might be going full circle here!]

**Comment by vext01** [comment]
So I think you are proposing just a renaming?

I'm happy with `start_tracing()` -> `start_recording()` and the same for stop, but it feels a bit odd for a `Trace` to record itself, so unless you feel strongly, I think the thing that does the recording should remain a `Tracer`.

> [I appreciate we might be going full circle here!]

It happens. Don't worry about it.

**Comment by ltratt** [comment]
I think the struct should be `Trace`, because that's what we care about (who cares about the tracer? we want the trace!) and it more nicely fits the multi-threaded idea. It means once I've done `start`/`stop`, I have a trace, which I can then do with what I want (store it, analyse it, discard it). I don't mind if the function names are`start/stop_tracing` or `record/stop` or whatever.

**Comment by vext01** [comment]
Hrm, I'm finding this problematic.

One instance of what is now called Tracer can make many traces. So just as the machine that can make many slices of toast is a toaster, the thing the makes the traces is called a tracer.

I think I'm going to be forever confused if we call this `Trace`. It's not a trace, it's a thing which makes traces.

Would it help if `stop_tracing()` returned a `trace`, rather than storing it? Just as my toaster returns toast?

**Comment by ltratt** [comment]
I have no idea what #23 means -- I assume that's a note to yourself ;)

I'm confused. If `Tracer` makes a trace *and* stores a trace, what happens if I want to either a) keep an old trace lurking around b) do two traces in parallel? It seems to me that 2 things are being conflated that shouldn't be.

That said, I don't mind if we end up with eithe:

```rust
let tracer = Tracer::new();
let trace = tracer.start_trace();
tracer.stop_trace(&trace);
```

or:

```rust
let tracer = Tracer::new();
let trace = tracer.recorder();
trace.stop();
```

or even:

```rust
let tracer = Tracer::new();
let trace = tracer.new_trace();
trace.record();
trace.stop();
```

All are basically morally equivalent in my book. I also don't feel strongly as to whether a `Trace` should be reusable, or a one-shot thing.

**Comment by vext01** [comment]
> I have no idea what #23 means

It's the issue to make the tracer instances re-usable. Since there's no point in re-opening perf file descriptors if we want to trace the same TID again.

> I'm confused. If Tracer makes a trace and stores a trace, what happens if I want to either a) keep an old trace lurking around.

Right now you would need to keep the tracer instance also lurking around, which isn't ideal, and hence my above comment: "Would it help if stop_tracing() returned a trace, rather than storing it?". And I think the answer is yes. This would mean a `Tracer` really is a machine that spits out traces.

> do two traces in parallel?

This should be possible. As we have established, each `Tracer` instance is unique to a TID. You should have no issues tracing more than one TID at a time with multiple instances of a `Tracer` (unless there is a hardware or perf constraint I'm unaware of -- there will be tests, don't worry!).

The interface I'm proposing above is more like:

```
let tracer = PerfPTTracer::new().unwrap();
tracer.start();
let trace: HWTrace = tracer.stop().unwrap();

// this part is subject to change, but for illustration purposes.
for jump in trace.jumps() {
   ...
}
```

or getting the trace out could be explicit, like:

```
let tracer = PerfPTTracer::new().unwrap();
tracer.start();
tracer.stop().unwrap();
let trace: HWTrace = tracer.trace();
```

So here, stopping the trace spits out a new struct which represents and "stores" a trace. I say "store" because all it really has inside is a pointer to some memory that was heap allocd in C.

The first and third examples in your last post were not quite how I imagined it being, but would work, the second I don't like so much due to `recorder`.

**Comment by ltratt** [comment]
I'm fine with:

```rust
let tracer = PerfPTTracer::new().unwrap();
tracer.start();
let trace: HWTrace = tracer.stop().unwrap();
```

[Although I'm always nervous about returning `Result` in too many places, I can imagine it might be necessary here.]

**Comment by vext01** [comment]
Great! I'll rework my code today :man_technologist: 

> I'm always nervous about returning Result in too many places

Well we can't get around the fact that just about every syscall I use underneath may fail. Realistically, we have two options:

 * panic
 * return a result

What the in-tree code does right now is the latter:
```
fn stop_tracing(&mut self) -> Result<(), TraceMeError>
```

So, if `stop_tracing` is going to make a trace, we would have:
```
fn stop_tracing(&mut self) -> Result<HWTrace, TraceMeError>
```

[i.e. the `Ok` wraps the trace.]

**Comment by ltratt** [comment]
Hmm, is there really one type for all `TraceMeError`s? That feels a bit weird to me, because I imagine the errors you can get when starting a trace are different than those when you end a trace. That said, we can always change the error type later, normally without hugely inconveniencing people.

**Comment by vext01** [comment]
There are many variants of `TraceMeError`. I was led to believe by the Rust docs that this is the Rust way, but I'm guessing you have another opinion?

https://github.com/softdevteam/traceme/blob/7d2adf27bda255a3045b9760c58945ab5dffbc13/src/errors.rs#L42-L56

Either way, we are getting lots of different aspects mixed up. If you think the way the errors are implemented is wrong, please raise an issue and we can look at that in another PR.

**Comment by vext01** [comment]
Fixed in #26
